File: node/src/OpenTelemetry.ts
/**
 * Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0
 */

import {
    ConfigurationError,
    InitOpenTelemetry,
    Logger,
    OpenTelemetryConfig,
} from ".";

/**
 * Extended OpenTelemetry configuration that includes JS-side options
 * not representable in the native (NAPI) config.
 *
 * Pass this to {@link OpenTelemetry.init} instead of the plain `OpenTelemetryConfig`.
 */
export interface GlideOpenTelemetryConfig extends OpenTelemetryConfig {
    /**
     * Optional callback that returns the active parent span context for each command.
     *
     * When a {@link GlideSpanContext} is returned, the GLIDE command span will be created
     * as a child of that context, enabling end-to-end distributed tracing.
     *
     * The callback is invoked synchronously before each sampled command. Keep the
     * implementation lightweight — avoid I/O, async work, or expensive computation.
     *
     * @example
     * ```typescript
     * import { trace } from "@opentelemetry/api";
     *
     * OpenTelemetry.init({
     *     traces: { endpoint: "http://localhost:4318/v1/traces" },
     *     parentSpanContextProvider: () => {
     *         const span = trace.getActiveSpan();
     *         if (!span) return undefined;
     *         const ctx = span.spanContext();
     *         return {
     *             traceId: ctx.traceId,
     *             spanId: ctx.spanId,
     *             traceFlags: ctx.traceFlags,
     *             traceState: ctx.traceState?.toString(),
     *         };
     *     },
     * });
     * ```
     */
    parentSpanContextProvider?: () => GlideSpanContext | undefined;
}

/**
 * Represents the trace context of a remote span, used for parent span context propagation.
 *
 * When a user's application has an active OTel span (e.g., from an HTTP request handler),
 * this context allows GLIDE command spans to appear as children of that span in tracing UIs.
 */
export interface GlideSpanContext {
    /** The trace ID as a 32-character lowercase hex string. */
    traceId: string;
    /** The span ID as a 16-character lowercase hex string. */
    spanId: string;
    /** Trace flags (e.g., 1 for sampled). */
    traceFlags: number;
    /** Optional W3C trace state (e.g., "vendorname1=opaqueValue1,vendorname2=opaqueValue2"). */
    traceState?: string;
}

/**
 * ⚠️ OpenTelemetry can only be initialized once per process. Calling `OpenTelemetry.init()` more than once will be ignored.
 * If you need to change configuration, restart the process with new settings.
 * ### OpenTelemetry
 *
 * - **openTelemetryConfig**: Use {@link GlideOpenTelemetryConfig} to configure OpenTelemetry exporters and options.
 *   - **traces**: (optional) Configure trace exporting.
 *     - **endpoint**: The collector endpoint for traces. Supported protocols:
 *       - `http://` or `https://` for HTTP/HTTPS
 *       - `grpc://` for gRPC
 *       - `file://` for local file export (see below)
 *     - **samplePercentage**: (optional) The percentage of requests to sample and create a span for, used to measure command duration. Must be between 0 and 100. Defaults to 1 if not specified.
 *       Note: There is a tradeoff between sampling percentage and performance. Higher sampling percentages will provide more detailed telemetry data but will impact performance.
 *       It is recommended to keep this number low (1-5%) in production environments unless you have specific needs for higher sampling rates.
 *   - **metrics**: (optional) Configure metrics exporting.
 *     - **endpoint**: The collector endpoint for metrics. Same protocol rules as above.
 *   - **flushIntervalMs**: (optional) Interval in milliseconds for flushing data to the collector. Must be a positive integer. Defaults to 5000ms if not specified.
 *   - **parentSpanContextProvider**: (optional) Callback returning the active parent span context. See {@link GlideOpenTelemetryConfig.parentSpanContextProvider}.
 *
 * #### File Exporter Details
 * - For `file://` endpoints:
 *   - The path must start with `file://` (e.g., `file:///tmp/otel` or `file:///tmp/otel/traces.json`).
 *   - If the path is a directory or lacks a file extension, data is written to `signals.json` in that directory.
 *   - If the path includes a filename with an extension, that file is used as-is.
 *   - The parent directory must already exist; otherwise, initialization will fail with an InvalidInput error.
 *   - If the target file exists, new data is appended (not overwritten).
 *
 * #### Validation Rules
 * - `flushIntervalMs` must be a positive integer.
 * - `samplePercentage` must be between 0 and 100.
 * - File exporter paths must start with `file://` and have an existing parent directory.
 * - Invalid configuration will throw an error synchronously when calling `OpenTelemetry.init()`.
 */
export class OpenTelemetry {
    private static _instance: OpenTelemetry | null = null;
    private static openTelemetryConfig: OpenTelemetryConfig | null = null;
    private static spanContextFn: (() => GlideSpanContext | undefined) | null =
        null;
    private static readonly TRACE_ID_REGEX = /^[0-9a-f]{32}$/;
    private static readonly SPAN_ID_REGEX = /^[0-9a-f]{16}$/;

    /**
     * Singleton class for managing OpenTelemetry configuration and operations.
     * This class provides a centralized way to initialize OpenTelemetry and control
     * sampling behavior at runtime.
     *
     * Example usage:
     * ```typescript
     * import { OpenTelemetry, GlideOpenTelemetryConfig } from "@valkey/valkey-glide";
     * import { trace } from "@opentelemetry/api";
     *
     * const config: GlideOpenTelemetryConfig = {
     *     traces: {
     *         endpoint: "http://localhost:4318/v1/traces",
     *         samplePercentage: 10,
     *     },
     *     metrics: {
     *         endpoint: "http://localhost:4318/v1/metrics",
     *     },
     *     flushIntervalMs: 1000,
     *     parentSpanContextProvider: () => {
     *         const span = trace.getActiveSpan();
     *         if (!span) return undefined;
     *         const ctx = span.spanContext();
     *         return { traceId: ctx.traceId, spanId: ctx.spanId, traceFlags: ctx.traceFlags };
     *     },
     * };
     * OpenTelemetry.init(config);
     * ```
     *
     * @remarks
     *   OpenTelemetry can only be initialized once per process. Subsequent calls to
     *   init() will be ignored. This is by design, as OpenTelemetry is a global
     *   resource that should be configured once at application startup.
     *
     * Initialize the OpenTelemetry instance
     * @param openTelemetryConfig - The OpenTelemetry configuration
     */
    public static init(openTelemetryConfig: GlideOpenTelemetryConfig) {
        if (!this._instance) {
            const { parentSpanContextProvider, ...nativeConfig } =
                openTelemetryConfig;
            this.internalInit(nativeConfig, parentSpanContextProvider);
            Logger.log(
                "info",
                "GlideOpenTelemetry",
                "OpenTelemetry initialized with config: " +
                    JSON.stringify(nativeConfig) +
                    (parentSpanContextProvider
                        ? " (parentSpanContextProvider: set)"
                        : ""),
            );
            return;
        }

        Logger.log(
            "warn",
            "GlideOpenTelemetry",
            "OpenTelemetry already initialized - ignoring new configuration",
        );
    }

    private static internalInit(
        nativeConfig: OpenTelemetryConfig,
        parentSpanContextProvider?: () => GlideSpanContext | undefined,
    ) {
        this.openTelemetryConfig = nativeConfig;

        if (parentSpanContextProvider) {
            this.spanContextFn = parentSpanContextProvider;
        }

        InitOpenTelemetry(nativeConfig);
        this._instance = new OpenTelemetry();
    }

    /**
     * Check if the OpenTelemetry instance is initialized
     * @returns True if the OpenTelemetry instance is initialized, false otherwise
     */
    public static isInitialized() {
        return this._instance != null;
    }

    /**
     * Get the sample percentage for traces
     * @returns The sample percentage for traces only if OpenTelemetry is initialized and the traces config is set, otherwise undefined.
     */
    public static getSamplePercentage() {
        return this.openTelemetryConfig?.traces?.samplePercentage;
    }

    /**
     * Determines if the current request should be sampled for OpenTelemetry tracing.
     * Uses the configured sample percentage to randomly decide whether to create a span for this request.
     * @returns true if the request should be sampled, false otherwise
     */
    public static shouldSample(): boolean {
        const percentage = this.getSamplePercentage();
        return (
            this.isInitialized() &&
            percentage !== undefined &&
            Math.random() * 100 < percentage
        );
    }

    /**
     * Set the percentage of requests to be sampled and traced. Must be a value between 0 and 100.
     * This setting only affects traces, not metrics.
     * @param percentage - The sample percentage 0-100
     * @throws Error if OpenTelemetry is not initialized or traces config is not set
     * @remarks
     * This method can be called at runtime to change the sampling percentage without reinitializing OpenTelemetry.
     */
    public static setSamplePercentage(percentage: number) {
        if (!this.openTelemetryConfig || !this.openTelemetryConfig.traces) {
            throw new ConfigurationError(
                "OpenTelemetry config traces not initialized",
            );
        }

        if (percentage < 0 || percentage > 100) {
            throw new ConfigurationError(
                "Sample percentage must be between 0 and 100",
            );
        }

        this.openTelemetryConfig.traces.samplePercentage = percentage;
    }

    /**
     * Register or replace the callback that returns the active parent span context.
     *
     * This allows changing the provider at runtime (e.g., switching tracing contexts
     * in a multi-tenant application). The initial provider can also be set via
     * {@link GlideOpenTelemetryConfig.parentSpanContextProvider} in `init()`.
     *
     * @param fn - A function returning a `GlideSpanContext` or `undefined`, or `null` to clear.
     *
     * @example
     * ```typescript
     * import { trace } from "@opentelemetry/api";
     *
     * OpenTelemetry.setParentSpanContextProvider(() => {
     *     const span = trace.getActiveSpan();
     *     if (!span) return undefined;
     *     const ctx = span.spanContext();
     *     return {
     *         traceId: ctx.traceId,
     *         spanId: ctx.spanId,
     *         traceFlags: ctx.traceFlags,
     *         traceState: ctx.traceState?.toString(),
     *     };
     * });
     * ```
     */
    public static setParentSpanContextProvider(
        fn: (() => GlideSpanContext | undefined) | null,
    ) {
        this.spanContextFn = fn;
    }

    /**
     * Retrieve the current parent span context by invoking the registered callback.
     *
     * @returns The `GlideSpanContext` from the registered callback, or `undefined` if no callback
     *   is set or the callback returns `undefined`.
     * @internal
     */
    public static getParentSpanContext(): GlideSpanContext | undefined {
        let ctx: GlideSpanContext | undefined;

        try {
            ctx = this.spanContextFn?.();
        } catch (e) {
            Logger.log(
                "warn",
                "GlideOpenTelemetry",
                `parentSpanContextProvider threw: ${e}. Falling back to standalone span.`,
            );
            return undefined;
        }

        if (ctx === undefined) {
            return undefined;
        }

        if (!this.TRACE_ID_REGEX.test(ctx.traceId)) {
            Logger.log(
                "warn",
                "GlideOpenTelemetry",
                `Invalid traceId "${ctx.traceId}" — expected 32 lowercase hex chars. Falling back to standalone span.`,
            );
            return undefined;
        }

        if (!this.SPAN_ID_REGEX.test(ctx.spanId)) {
            Logger.log(
                "warn",
                "GlideOpenTelemetry",
                `Invalid spanId "${ctx.spanId}" — expected 16 lowercase hex chars. Falling back to standalone span.`,
            );
            return undefined;
        }

        if (
            !Number.isInteger(ctx.traceFlags) ||
            ctx.traceFlags < 0 ||
            ctx.traceFlags > 255
        ) {
            Logger.log(
                "warn",
                "GlideOpenTelemetry",
                `Invalid traceFlags "${ctx.traceFlags}" — expected integer 0-255. Falling back to standalone span.`,
            );
            return undefined;
        }

        if (
            ctx.traceState !== undefined &&
            !OpenTelemetry.validTraceState(ctx.traceState)
        ) {
            Logger.log(
                "warn",
                "GlideOpenTelemetry",
                `Invalid traceState "${ctx.traceState}" — expected W3C tracestate format. Falling back to standalone span.`,
            );
            return undefined;
        }

        return ctx;
    }

    /**
     * Validate a W3C tracestate key.
     * See https://www.w3.org/TR/trace-context/#key
     * @internal
     */
    private static validTraceStateKey(key: string): boolean {
        if (key.length === 0 || key.length > 256) return false;

        const ALLOWED_SPECIAL = new Set([
            "_".charCodeAt(0),
            "-".charCodeAt(0),
            "*".charCodeAt(0),
            "/".charCodeAt(0),
        ]);

        let vendorStart: number | null = null;

        for (let i = 0; i < key.length; i++) {
            const c = key.charCodeAt(i);
            const isLower = c >= 0x61 && c <= 0x7a; // a-z
            const isDigit = c >= 0x30 && c <= 0x39; // 0-9
            const isAt = c === 0x40; // @

            if (!(isLower || isDigit || ALLOWED_SPECIAL.has(c) || isAt)) {
                return false;
            }

            if (i === 0 && !isLower && !isDigit) return false;

            if (isAt) {
                if (vendorStart !== null || i + 14 < key.length) return false;
                vendorStart = i;
            } else if (vendorStart !== null && i === vendorStart + 1) {
                if (!isLower && !isDigit) return false;
            }
        }

        return true;
    }

    /**
     * Validate a W3C tracestate value.
     * See https://www.w3.org/TR/trace-context/#value
     * @internal
     */
    private static validTraceStateValue(value: string): boolean {
        if (value.length > 256) return false;

        return !value.includes(",") && !value.includes("=");
    }

    /**
     * Validate a W3C tracestate string (comma-separated key=value pairs).
     * Mirrors the validation in opentelemetry-rust TraceState::from_str.
     * @internal
     */
    private static validTraceState(traceState: string): boolean {
        if (traceState === "") return true;

        const entries = traceState.split(",");

        for (const entry of entries) {
            const eqIndex = entry.indexOf("=");

            if (eqIndex === -1) return false;

            const key = entry.substring(0, eqIndex);
            const value = entry.substring(eqIndex + 1);

            if (
                !OpenTelemetry.validTraceStateKey(key) ||
                !OpenTelemetry.validTraceStateValue(value)
            ) {
                return false;
            }
        }

        return true;
    }
}

File: python/glide-shared/glide_shared/opentelemetry.py
# Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

"""
OpenTelemetry configuration classes shared between async and sync clients.

OpenTelemetry can only be initialized once per process. Calling OpenTelemetry.init()
more than once will be ignored. If you need to change configuration, restart the process
with new settings.

OpenTelemetry Configuration
----------------------------

OpenTelemetryConfig: Main configuration object for OpenTelemetry exporters and options.

* traces: (optional) Configure trace exporting using OpenTelemetryTracesConfig.

  * endpoint: The collector endpoint for traces. Supported protocols:
    http://, https:// for HTTP/HTTPS, grpc:// for gRPC, file:// for local file export
  * sample_percentage: (optional) The percentage of requests to sample (0-100). Defaults to 1.
    Note: Higher sampling percentages impact performance. Recommended: 1-5% in production.

* metrics: (optional) Configure metrics exporting using OpenTelemetryMetricsConfig.

  * endpoint: The collector endpoint for metrics. Same protocol rules as above.

* flush_interval_ms: (optional) Interval in milliseconds for flushing data. Defaults to 5000ms.

File Exporter Details
---------------------

For file:// endpoints:

* Path must start with file:// (e.g., file:///tmp/otel or file:///tmp/otel/traces.json)
* If path is a directory or lacks extension, data is written to signals.json in that directory
* If path includes filename with extension, that file is used as-is
* Parent directory must exist; otherwise initialization fails with InvalidInput error
* If target file exists, new data is appended (not overwritten)

Validation Rules
----------------

* flush_interval_ms must be a positive integer
* sample_percentage must be between 0 and 100
* File exporter paths must start with file:// and have an existing parent directory
* Invalid configuration will throw an error when calling OpenTelemetry.init()

Trace Context Propagation
-------------------------

When the application has its own active OpenTelemetry span, GLIDE creates its command,
batch and script spans as children of it, so a trace has no gap at the database
boundary. The parent is read from the OpenTelemetry Python API, which propagates the
active span through ``contextvars``, so nothing is passed to GLIDE. It requires the
optional ``opentelemetry-api`` package. Without it, propagation is off and spans are
created as independent trace roots.

See the "Trace Context Propagation" section of the Python README for the user-facing
description, including how ``sample_percentage`` interacts with the parent span and
where GLIDE's spans are exported.
"""

import random
from dataclasses import dataclass
from typing import Any, Optional, Tuple

from glide_shared.exceptions import ConfigurationError

try:
    from opentelemetry import trace as _otel_trace
except ImportError:
    _otel_trace = None  # type: ignore[assignment]


@dataclass
class _ParentSpanContext:
    """A valid parent span context extracted from the OpenTelemetry context."""

    __slots__ = ("trace_id", "span_id", "trace_flags", "trace_state")

    trace_id: bytes
    span_id: bytes
    trace_flags: int
    trace_state: Optional[bytes]

    def to_cstrings(self, ffi: Any) -> Tuple[Any, Any, Any]:
        """Marshal the trace context into C strings for an FFI span creation call.

        The returned ``char[]`` buffers must stay referenced for the duration of the
        FFI call, the caller keeps them alive by holding them until the call returns.

        Args:
            ffi: The CFFI ``ffi`` object from ``GlideFFI``.

        Returns:
            Tuple[Any, Any, Any]: The trace_id, span_id and trace_state buffers, the
                last being ``ffi.NULL`` when there is no tracestate.
        """
        return (
            ffi.new("char[]", self.trace_id),
            ffi.new("char[]", self.span_id),
            (
                ffi.new("char[]", self.trace_state)
                if self.trace_state is not None
                else ffi.NULL
            ),
        )


def _create_command_span(
    ffi: Any, lib: Any, span_name_cstr: Any, parent: Optional[_ParentSpanContext]
) -> int:
    """Create a command span, parented to ``parent`` when one was extracted.

    Args:
        ffi: The CFFI ``ffi`` object from ``GlideFFI``.
        lib: The CFFI ``lib`` object from ``GlideFFI``.
        span_name_cstr: A ``char[]`` holding the span name.
        parent: The extracted parent context, or ``None`` for an independent root span.

    Returns:
        int: The span pointer, or 0 if span creation failed.
    """
    if parent is None:
        return lib.create_named_otel_span(span_name_cstr)

    # The buffers stay referenced by these locals across the call below.
    trace_id_cstr, span_id_cstr, trace_state_cstr = parent.to_cstrings(ffi)
    return lib.create_named_otel_span_with_trace_context(
        span_name_cstr,
        trace_id_cstr,
        span_id_cstr,
        parent.trace_flags,
        trace_state_cstr,
    )


def _create_batch_span(ffi: Any, lib: Any, parent: Optional[_ParentSpanContext]) -> int:
    """Create a batch span, parented to ``parent`` when one was extracted.

    Args:
        ffi: The CFFI ``ffi`` object from ``GlideFFI``.
        lib: The CFFI ``lib`` object from ``GlideFFI``.
        parent: The extracted parent context, or ``None`` for an independent root span.

    Returns:
        int: The span pointer, or 0 if span creation failed.
    """
    if parent is None:
        return lib.create_batch_otel_span()

    # The buffers stay referenced by these locals across the call below.
    trace_id_cstr, span_id_cstr, trace_state_cstr = parent.to_cstrings(ffi)
    return lib.create_batch_otel_span_with_trace_context(
        trace_id_cstr,
        span_id_cstr,
        parent.trace_flags,
        trace_state_cstr,
    )


class OpenTelemetryTracesConfig:
    """Configuration for exporting OpenTelemetry traces."""

    def __init__(self, endpoint: str, sample_percentage: Optional[int] = None) -> None:
        self.endpoint = endpoint
        self.sample_percentage = (
            sample_percentage if sample_percentage is not None else 1
        )

    def get_endpoint(self) -> str:
        return self.endpoint

    def get_sample_percentage(self) -> int:
        return self.sample_percentage


class OpenTelemetryMetricsConfig:
    """Configuration for exporting OpenTelemetry metrics."""

    def __init__(self, endpoint: str) -> None:
        self.endpoint = endpoint

    def get_endpoint(self) -> str:
        return self.endpoint


class OpenTelemetryConfig:
    """Configuration for OpenTelemetry integration."""

    def __init__(
        self,
        traces: Optional[OpenTelemetryTracesConfig] = None,
        metrics: Optional[OpenTelemetryMetricsConfig] = None,
        flush_interval_ms: Optional[int] = None,
    ) -> None:
        self.traces = traces
        self.metrics = metrics
        self.flush_interval_ms = flush_interval_ms

    def get_traces(self) -> Optional[OpenTelemetryTracesConfig]:
        return self.traces

    def set_traces(self, traces: OpenTelemetryTracesConfig) -> None:
        self.traces = traces

    def get_metrics(self) -> Optional[OpenTelemetryMetricsConfig]:
        return self.metrics

    def get_flush_interval_ms(self) -> Optional[int]:
        return self.flush_interval_ms


class OpenTelemetry:
    """
    Singleton class for managing OpenTelemetry configuration and operations.

    This class provides a centralized way to initialize OpenTelemetry and control
    sampling behavior at runtime.

    Example usage::

        from glide import OpenTelemetry, OpenTelemetryConfig, OpenTelemetryTracesConfig, OpenTelemetryMetricsConfig

        OpenTelemetry.init(OpenTelemetryConfig(
            traces=OpenTelemetryTracesConfig(
                endpoint="http://localhost:4318/v1/traces",
                sample_percentage=10
            ),
            metrics=OpenTelemetryMetricsConfig(
                endpoint="http://localhost:4318/v1/metrics"
            ),
            flush_interval_ms=1000
        ))

    Note:
        OpenTelemetry can only be initialized once per process. Subsequent calls to
        init() will be ignored. This is by design, as OpenTelemetry is a global
        resource that should be configured once at application startup.
    """

    _instance: Optional["OpenTelemetry"] = None
    _config: Optional[OpenTelemetryConfig] = None

    @classmethod
    def init(cls, config: OpenTelemetryConfig) -> None:
        """Initialize OpenTelemetry with the given configuration.

        This method can only be called once per process. Subsequent calls will log
        a warning and be ignored.

        Args:
            config: The OpenTelemetry configuration specifying trace/metrics endpoints
                and flush intervals.

        Raises:
            ConfigurationError: If the underlying FFI initialization fails (e.g., invalid
                endpoint or file path).
        """
        if not cls._instance:
            cls._config = config
            from glide_shared._glide_ffi import GlideFFI
            from glide_shared.logger import Level, Logger

            ffi = GlideFFI.ffi
            lib = GlideFFI.lib

            traces_ptr = ffi.NULL
            traces_endpoint_cstr = None
            if config.traces:
                traces_endpoint_cstr = ffi.new(
                    "char[]", config.traces.endpoint.encode()
                )
                traces_config = ffi.new("OpenTelemetryTracesConfig*")
                traces_config.endpoint = traces_endpoint_cstr
                traces_config.has_sample_percentage = True
                traces_config.sample_percentage = config.traces.sample_percentage
                traces_ptr = traces_config

            metrics_ptr = ffi.NULL
            metrics_endpoint_cstr = None
            if config.metrics:
                metrics_endpoint_cstr = ffi.new(
                    "char[]", config.metrics.endpoint.encode()
                )
                metrics_config = ffi.new("OpenTelemetryMetricsConfig*")
                metrics_config.endpoint = metrics_endpoint_cstr
                metrics_ptr = metrics_config

            otel_config = ffi.new("OpenTelemetryConfig*")
            otel_config.traces = traces_ptr
            otel_config.metrics = metrics_ptr
            otel_config.has_flush_interval_ms = config.flush_interval_ms is not None
            otel_config.flush_interval_ms = (
                config.flush_interval_ms if config.flush_interval_ms else 0
            )

            error = lib.init_open_telemetry(otel_config)
            if error != ffi.NULL:
                error_msg = ffi.string(error).decode()
                lib.free_c_string(error)
                raise ConfigurationError(
                    f"Failed to initialize OpenTelemetry: {error_msg}"
                )

            cls._instance = OpenTelemetry()
            Logger.log(
                Level.INFO,
                "GlideOpenTelemetry",
                "OpenTelemetry initialized successfully",
            )
            return

        from glide_shared.logger import Level, Logger

        Logger.log(
            Level.WARN,
            "GlideOpenTelemetry",
            "OpenTelemetry already initialized - ignoring new configuration",
        )

    @classmethod
    def is_initialized(cls) -> bool:
        """Check whether OpenTelemetry has been initialized.

        Returns:
            bool: True if init() has been successfully called, False otherwise.
        """
        return cls._instance is not None

    @classmethod
    def is_tracing_enabled(cls) -> bool:
        """Check whether trace exporting was successfully initialized.

        Returns:
            bool: True if OpenTelemetry is initialized with traces configured.
        """
        return (
            cls._instance is not None
            and cls._config is not None
            and cls._config.traces is not None
        )

    @classmethod
    def get_sample_percentage(cls) -> Optional[int]:
        """Get the current trace sampling percentage.

        Returns:
            Optional[int]: The sampling percentage (0-100), or None if traces are not configured.
        """
        if cls._config and cls._config.traces:
            return cls._config.traces.sample_percentage
        return None

    @classmethod
    def should_sample(cls) -> bool:
        """Determine whether the current request should be sampled.

        Uses random sampling based on the configured sample percentage.

        Returns:
            bool: True if the request should be traced, False otherwise.
                Always returns False if OpenTelemetry is not initialized.
        """
        if not cls._instance:
            return False
        percentage = cls.get_sample_percentage()
        return percentage is not None and random.random() * 100 < percentage

    @classmethod
    def _get_parent_span_context(cls) -> Optional[_ParentSpanContext]:
        """Extract the OpenTelemetry span context, if it is usable as a parent.

        Returns:
            Optional[_ParentSpanContext]: The active span context, or None if
                ``opentelemetry-api`` is not installed, no valid span is active,
                or the context could not be read.
        """
        if _otel_trace is None:
            return None

        try:
            span_context = _otel_trace.get_current_span().get_span_context()
            if not span_context.is_valid:
                return None
            trace_flags = span_context.trace_flags

            # to_header() is the W3C `tracestate` serialization
            trace_state = span_context.trace_state
            trace_state_header = trace_state.to_header() if trace_state else ""
            return _ParentSpanContext(
                trace_id=format(span_context.trace_id, "032x").encode(),
                span_id=format(span_context.span_id, "016x").encode(),
                trace_flags=int(trace_flags),
                trace_state=trace_state_header.encode() or None,
            )
        except Exception as e:
            from glide_shared.logger import Level, Logger

            Logger.log(
                Level.DEBUG,
                "GlideOpenTelemetry",
                f"Failed to read the active span context: {e}. "
                "Continuing as if no span were active.",
            )
            return None

    @classmethod
    def set_sample_percentage(cls, percentage: int) -> None:
        """Update the trace sampling percentage at runtime.

        Args:
            percentage: The new sampling percentage (0-100).

        Raises:
            ConfigurationError: If traces are not initialized or percentage is out of range.
        """
        if not cls._config or not cls._config.traces:
            raise ConfigurationError("OpenTelemetry traces not initialized")
        if percentage < 0 or percentage > 100:
            raise ConfigurationError("Sample percentage must be between 0 and 100")
        cls._config.traces.sample_percentage = percentage
